components/normalizer/src/lib.rs

```diff
@@ -1438,32 +1438,40 @@ macro_rules! normalizer_methods {
     () => {
         /// Normalize a string slice into a `Cow<'a, str>`.
         pub fn normalize<'a>(&self, text: &'a str) -> Cow<'a, str> {
-            let up_to = self.is_normalized_up_to(text);
-            if up_to == text.len() {
-                return Cow::Borrowed(text);
+            let (head, tail) = self.split_normalized(text);
+            if tail.is_empty() {
+                return Cow::Borrowed(head);
             }
             let mut ret = String::new();
             ret.reserve(text.len());
-            let (head, tail) = text.split_at(up_to);
             ret.push_str(head);
             let _ = self.normalize_to(tail, &mut ret);
             Cow::Owned(ret)
         }
 
+        /// Split a string slice into maximum normalized prefix and unnormalized suffix
+        /// such that the concatenation of the prefix and the normalization of the suffix
+        /// is the normalization of the whole input.
+        pub fn split_normalized<'a>(&self, text: &'a str) -> (&'a str, &'a str) {
+            if let Some(pair) = text.split_at_checked(self.is_normalized_up_to(text)) {
```

Comment:
thought: a way to land this without changing MSRV is to explicitly perform the bounds check before calling `.split_at()`. Then in the future we can use the better API.